File: app/src/main/java/com/lalema/app/domain/PoopRepository.kt
package com.lalema.app.domain

import com.lalema.app.data.PoopRecord
import com.lalema.app.data.PoopRecordDao
import kotlinx.coroutines.flow.Flow
import java.time.LocalDate
import java.time.format.DateTimeFormatter
class PoopRepository(
    private val dao: PoopRecordDao
) {

    private val dateFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")

    suspend fun record(
        date: String,
        timeHour: Int = java.time.LocalTime.now().hour,
        timeMinute: Int = java.time.LocalTime.now().minute,
        amount: String = "NORMAL",
        consistency: String = "NORMAL",
        color: String = "BROWN",
        smell: String = "NORMAL",
        painLevel: Int = 0,
        blood: Boolean = false,
        mucus: Boolean = false,
        notes: String = ""
    ): Boolean {
        val result = dao.insert(
            PoopRecord.create(
                date = date,
                timeHour = timeHour,
                timeMinute = timeMinute,
                amount = amount,
                consistency = consistency,
                color = color,
                smell = smell,
                painLevel = painLevel,
                blood = blood,
                mucus = mucus,
                notes = notes
            )
        )
        return result != -1L
    }

    suspend fun isRecorded(date: String): Boolean {
        return dao.existsByDate(date)
    }

    suspend fun getByDate(date: String): List<PoopRecord> {
        return dao.getByDate(date)
    }

    suspend fun getByDateRange(startDate: String, endDate: String): List<PoopRecord> {
        return dao.getByDateRange(startDate, endDate)
    }

    fun getAllFlow(): Flow<List<PoopRecord>> {
        return dao.getAll()
    }

    suspend fun getAll(): List<PoopRecord> {
        return dao.getAllList()
    }

    suspend fun getCountByMonth(year: Int, month: Int): Int {
        val pattern = String.format("%04d-%02d-%%", year, month)
        return dao.getCountByMonth(pattern)
    }

    suspend fun getRecordCountByMonth(year: Int, month: Int): Int {
        val pattern = String.format("%04d-%02d-%%", year, month)
        return dao.getRecordCountByMonth(pattern)
    }

    suspend fun getStreak(): Int {
        var streak = 0
        var current = LocalDate.now()
        var daysChecked = 0
        while (daysChecked < 365) {
            val dateStr = current.format(dateFormatter)
            if (dao.existsByDate(dateStr)) {
                streak++
                current = current.minusDays(1)
                daysChecked++
            } else {
                break
            }
        }
        return streak
    }

    suspend fun deleteByDate(date: String) {
        dao.deleteByDate(date)
    }

    suspend fun deleteRecord(id: Long) {
        dao.deleteById(id)
    }
}

File: LaLeMa-iOS/Sources/Services/PoopRepository.swift
import Foundation

final class PoopRepository {
    static let shared = PoopRepository()

    private let database = DatabaseManager.shared
    private let dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter
    }()

    private init() {}

    func record(
        date: String,
        timeHour: Int,
        timeMinute: Int,
        amount: PoopAmount,
        consistency: PoopConsistency,
        color: PoopColor,
        smell: PoopSmell,
        painLevel: Int,
        blood: Bool,
        mucus: Bool,
        notes: String
    ) -> Bool {
        let record = PoopRecord(
            date: date,
            timeHour: timeHour,
            timeMinute: timeMinute,
            amount: amount,
            consistency: consistency,
            color: color,
            smell: smell,
            painLevel: painLevel,
            blood: blood,
            mucus: mucus,
            notes: notes
        )

        if let _ = database.insertRecord(record) {
            return true
        }
        return false
    }

    func isRecorded(date: String) -> Bool {
        return database.existsByDate(date)
    }

    func getByDate(_ date: String) -> [PoopRecord] {
        return database.getRecordsByDate(date)
    }

    func getByDateRange(startDate: String, endDate: String) -> [PoopRecord] {
        return database.getRecordsByDateRange(startDate: startDate, endDate: endDate)
    }

    func getCountByMonth(year: Int, month: Int) -> Int {
        return database.getDistinctDateCountByMonth(year: year, month: month)
    }

    func getRecordCountByMonth(year: Int, month: Int) -> Int {
        return database.getRecordCountByMonth(year: year, month: month)
    }

    func getStreak() -> Int {
        var streak = 0
        var currentDate = Date()
        var daysChecked = 0

        while daysChecked < 365 {
            let dateString = dateFormatter.string(from: currentDate)
            if database.existsByDate(dateString) {
                streak += 1
                currentDate = Calendar.current.date(byAdding: .day, value: -1, to: currentDate) ?? currentDate
                daysChecked += 1
            } else {
                break
            }
        }

        return streak
    }

    func deleteByDate(_ date: String) {
        database.deleteByDate(date)
    }

    func deleteRecord(id: Int64) {
        database.deleteRecordById(id)
    }
}
